jq/exec.go

```diff
@@ -0,0 +1,314 @@
+)
+
+func errorFromJV(jv C.jv) error {
+	jv = C.jq_format_error(jv)
+	defer C.jv_free(jv)
+
+	if C.jv_get_kind(jv) == C.JV_KIND_NULL {
+		return nil
+	}
+	return errors.New(C.GoString(C.jv_string_value(jv)))
+}
+
+//export goJQErrorHandler
+func goJQErrorHandler(id uint64, jv C.jv) {
+	callbackErrorsM.Lock()
+	defer callbackErrorsM.Unlock()
+
+	err := errorFromJV(jv)
+	if err == nil {
+		return
+	}
+
+	callbackErrors[id] <- err
+}
+
+// Exec compiles a JQ program with the provided args and executes it with the
+// provided input.
+//
+// If the args parameter is not nil, a slice/array, or map[string]interface{},
+// then ErrWrongType is returned.
+func Exec(program string, args, input interface{}) ([]string, error) {
+	state, err := C.jq_init()
+	if err != nil {
+		return nil, err
+	} else if state == nil {
+		panic("failed to initialize jq state")
+	}
+	defer C.jq_teardown(&state)
+
+	argsJV := jvInterface(args)
+	defer C.jv_free(argsJV)
+
+	inputJV := jvInterface(input)
+	defer C.jv_free(inputJV)
+
+	return executeProgram(state, program, argsJV, inputJV)
+}
+
+// executeProgram compiles and executes a jq program with the provided
+// arguments and input.
+func executeProgram(state *C.struct_jq_state, program string, args, input C.jv) ([]string, error) {
+	errs := compile(state, program, args)
+	if len(errs) != 0 {
+		err := errs[0]
+		for i := 1; i < len(errs); i++ {
+			err = errors.New(err.Error() + "; " + errs[i].Error())
+		}
+		return nil, err
+	}
+
+	return execute(state, input)
+}
+
+// execute performs an execution of the previous compiled program.
+// compile() must be called before this function.
+func execute(state *C.struct_jq_state, input C.jv) ([]string, error) {
+	// I can't figure out where, but it seems like jq_start frees input.
+	C.jq_start(state, C.jv_copy(input), C.int(0))
+
+	results := make([]string, 0)
+	result := C.jq_next(state)
+	for C.jv_is_valid(result) == 1 {
+		results = append(results, dumpJvToGoStr(result))
+		C.jv_free(result)
+		result = C.jq_next(state)
+	}
+	defer C.jv_free(result)
+
+	return results, invalidError(result)
+}
+
+func invalidError(jv C.jv) error {
+	// jv_invalid_get_msg frees jv.
+	msg := C.jv_invalid_get_msg(jv)
+	defer C.jv_free(msg)
+
+	switch C.jv_get_kind(msg) {
+	case C.JV_KIND_NULL:
+		return nil
+	case C.JV_KIND_STRING:
+		return errors.New(C.GoString(C.jv_string_value(msg)))
+	default:
+		return errors.New(dumpJvToGoStr(msg))
+	}
+}
+
+// collectErrors wraps a closure that calls jv functions that perform error
+// handling via callback.
+func collectErrors(state *C.struct_jq_state, fn func()) []error {
+	callbackKey := rand.Uint64()
+	C.register_jq_error_cb(state, C.ulonglong(callbackKey))
+	defer C.set_jq_error_cb_default(state)
+
+	errChan := make(chan error)
+	callbackErrorsM.Lock()
+	callbackErrors[callbackKey] = errChan
+	callbackErrorsM.Unlock()
+
+	defer func() {
+		callbackErrorsM.Lock()
+		delete(callbackErrors, callbackKey)
+		callbackErrorsM.Unlock()
+	}()
+
+	go func() {
```

Comment:
There isn't any code ensuring this doesn't leak. You might want to use a waitGroup to ensure it exits before this function returns.

Reply:
Ranging over the channel should loop forever until we get nil.

Reply:
Might add a comment indicating this